File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.java
class Solution {
    public long subArrayRanges(int[] nums) {
        Stack<Integer> stk = new Stack<>();
        long sum = 0;
        int n = nums.length;
        for(int i=0; i<=n; i++){
            while(!stk.empty() && nums[stk.peek()] < (i==n?Integer.MAX_VALUE:nums[i])) {
                int j = stk.pop();
                int k = stk.empty() ? -1 : stk.peek();
                sum += (long)nums[j]*((j-k)*(i-j));
            }
            stk.push(i);
        }
        stk.clear();
        for(int i=0; i<=n; i++){
            while(!stk.empty() && nums[stk.peek()] > (i==n?Integer.MIN_VALUE:nums[i])) {
                int j = stk.pop();
                int k = stk.empty() ? -1 : stk.peek();
                sum -= (long)nums[j]*((j-k)*(i-j));
            }
            stk.push(i);
        }
        return sum;
    }
}

File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.py
class Solution:
    def subArrayRanges(self, nums: List[int]) -> int:
        n = len(nums)
        res = 0
        stack = []
        for right in range(n+1):
            while stack and (right==n or nums[stack[-1]]>=nums[right]):
                mid = stack.pop()
                left = stack[-1] if stack else -1
                res -= nums[mid]*(mid-left)*(right-mid)
            stack.append(right)
        stack.clear()
        for right in range(n+1):
            while stack and (right==n or nums[stack[-1]]<=nums[right]):
                mid = stack.pop()
                left = stack[-1] if stack else -1
                res += nums[mid]*(mid-left)*(right-mid)
            stack.append(right)
        return res